### ISSUE TITLE
feature: Use tcmalloc in supported platform as default memory allocator

### ISSUE THREAD

**Issue by Zheaoli** [bentoml/BentoML#4766]
### Feature request

Related to #4760 

Using tcmalloc will help us to avoid memory leaks when memory fragmentation existed 

TL;DR;

The glibc malloc is the most popular memory allocator in most distributions of Linux. We use `malloc` to allocate memory from the system and `free` to return the memory to the system. But there is a problem. When we call `free` , the memory may not return to system immediately. The glibc malloc will cache the small memory into a buffer and return it into system when it become a big chunk. But if the memory we allocate from the system is too small and too split to merge into a big chunk, the memory will be in used status for a long time even if we have already return it. We call it memory leak.

In Python, the developer warp the glibc allocator use the `munmap` API to ensure the memory returned into system immediately. FYI https://github.com/python/cpython/blob/main/Objects/obmalloc.c#L380-L395. So the memory leak is fine when our project is pure Python code

But we always use a lot of C extensions for Python. The extension may not use the Python memory allocator and use `malloc` and `free` API directly. So the memory leak again.

FYI https://github.com/protocolbuffers/protobuf/issues/10088#issuecomment-1587627538

So we have two way to solve this

1. Binding glibc and run `malloc_trim` background periodically
2. Use `tcmalloc` instead of the glibc malloc

I perfer 2
